### ISSUE TITLE
Store inconsistency when using selectors

### ISSUE THREAD

**Issue by maxime1992** [ngrx/store#312]
Hi, I'm using selectors a lot into an application built with ngrx and I have a problem with my store, **inside my selectors**.

I made a Plunkr to demonstrate my problem : https://plnkr.co/edit/9aOGse?p=preview
(it has many comments to give as much details as possible, just take a look to the code).

But let me explain here what's going on : 
## Data
The small app is composed of 2 reducers : `UsersReducer` & `PetsReducer`.
It's a *one to many relation* : A user can have many pets.
I normalized the data and our reducers manipulate those types : `IUsers` and `IPets`.

```
// USERS
export interface IUser {
  id: string;
  name: string;
  pets: string[];
}

export interface IUsers {
  isFetchingUsers: boolean;
  
  byId: { [key: string]: IUser };
  allIds: string[];
}

// PETS
export interface IPet {
  id: string;
  name: string;
}

export interface IPets {
  byId: { [key: string]: IPet };
  allIds: string[];
}
```

## How to trigger the "bug"
At the beginning we have the following :  
![image](https://cloud.githubusercontent.com/assets/4950209/21892514/7f959134-d8d7-11e6-8e83-dbf401ef3f4f.png)

Our selector (last one displayed doesn't have any value as it's supposed to be an array of users (containing pets) but right now we don't have any user.

When clicking on the button, here's what's happening : 
- dispatch an action `FETCH_USERS`
- an effect catches the `FETCH_USERS` action and call a (mock) service (to simulate async response)
- this effect then dispatch an action `FETCH_USERS_SUCCESS` handled by **both reducers** : `UsersReducer` and `PetsReducer`.

The selector throws an error : 
![image](https://cloud.githubusercontent.com/assets/4950209/21892683/461f58a8-d8d8-11e6-9c7b-6e047eaf83d6.png)
(we'll get back to that)

## what's happening ?
I put some console.log/debug in the selector to understand and here's the result : 
![image](https://cloud.githubusercontent.com/assets/4950209/21892962/690f78b0-d8d9-11e6-8f38-bcd1fefabe29.png)

- 1 | the selector is triggered before the action is even dispatched to any of the reducers : Consistent state --> 0 Users, 0 Pets
- 2 | the selector is triggered just after the action has been dispatched to the first reducer (`UsersReducer`) : **Inconsistent state** --> 2 Users (having a list of pets), 0 Pets

When the error is thrown, we can see that our action is even **NOT** in the debug panel : 
![image](https://cloud.githubusercontent.com/assets/4950209/21893105/f5963e68-d8d9-11e6-8fb9-ffb42fb63f4f.png)
So the action wasn't fully dispatched and our selectors are already kickin' in ?

I started to question myself and I thought that maybe I didn't get how Redux was working.
That maybe when an action is handled by multiple reducers, we would be notified for **every reducer done applying the action** and not at the end of the action, applied on every reducers.

So I went on Gitter (in the Redux chanel) and @rscarvalho told me that : 
![image](https://cloud.githubusercontent.com/assets/4950209/21893267/b3bfa1fe-d8da-11e6-8573-2acc631a9811.png)

So I don't know if I'm missing something in my selector or if there's a bug in ngrx, but I'm kinda lost here.

Any help would be nice.

Thanks !

PS : I'd like to thanks @rscarvalho, @jordic and @adharris for digging into that and helping me on Gitter :+1:

**Comment by maxime1992** [comment]
I started to dig into code source and I try to understand : 
https://github.com/ngrx/store/blob/master/src/state.ts

![image](https://cloud.githubusercontent.com/assets/4950209/21894791/bd23120c-d8e0-11e6-9da9-7f6c90209097.png)

I thought it'd be a `reduce` and I found a `scan`.
Looking at the doc 
Scan emits every value when they changes
![image](https://cloud.githubusercontent.com/assets/4950209/21894980/68f0db0a-d8e1-11e6-9901-3d8760400b35.png)

Reduce emits only one value at the end
![image](https://cloud.githubusercontent.com/assets/4950209/21894834/ebeefdee-d8e0-11e6-8c58-39119dbba7bb.png)

Could it be from there ?
Not sure if this function is for handling one action on every reducer or handling every actions over time ...

**Comment by abdulhaq-e** [comment]
I have not spent much time reading the plnkr but my first guess would be: it's quite normal to check for `undefined` variables in selectors before using it or accessing any property stored in it.

```ts
          {
            pets: users.byId[userId].pets.map(petId => {
              if (petId) {
              console.log(pets.byId[petId].name);
              return pets.byId[petId];
              } else {
              return {}; // or anything eles
              }
            })
          }
```

There are also other ways, like delaying the `selector` until the value is received.

**Comment by maxime1992** [comment]
@adharris showed me another solution which is working fine but I still think that the behavior I described previously is a problem.

```
export const getUsersWithPets: Observable<IUsersWithPets> = () =>
  (store) => store.select(s => [s.users, s.pets])
    // don't **have to** use distinctUntilChanged but it's way better for perf
    .distinctUntilChanged((a, b) => a[0] == b[0] && a[1] == b[1])
    .map(([users, pets]) => {
      console.group('Selector receiving values ...');
```

@abdulhaq-e I don't agree with you. I stick with the principle that data coming from the server can be trusted and that my store is consistent. Even if I were OK with that, triggering an event between each apply on the reducers of an action (the same action) doesn't seem right.

**Comment by abdulhaq-e** [comment]
I understand your problem very well, I have it as well. The solution I wrote is what I use. I'll be watching this for better solutions than mine.

I think the difference between `withLatestFrom` and `combineLatest` which is causing the latter to raise an error in your example is: `withLatestFrom` waits for ALL input observables to emit while `combineLatest` doesn't.

From docs, http://reactivex.io/rxjs/class/es6/Observable.js~Observable.html#instance-method-withLatestFrom :

> All input Observables must emit at least one value before the output Observable will emit a value.

I believe this doesn't apply for `combineLatest`. 

But as I understand, you still have two issues:
1- Why is the selector being triggered before dispatching any actions.
2- Why is the selector being triggered after only one reducer has been run.

**Comment by adharris** [comment]
My understanding is that this is happening because of the use of `combineLatest` in the selector:

```typescript
(store) => store.select(s => s.users)
    .combineLatest(store.select(s => s.pets))
    .map(([users, pets]) => {
```

So I can reference things by name, i'll extract the observables:

```typescript
let users$ = store.select(s => s.users);
let pets$  =store.select(s => s.pets);
return $users.combineLatest(pets$).map(([users, pets]) => ...
```

Combine latest works by caching the most recent emitted value of each observable, so that it when a new value comes in on one observable, it itself can emit both the new value and the old value for the second observable.

so for users and pets, combine latest will hang onto the most recent `users` and the most recent `pets`.  When `users$` emits a new value combine latest will emit a new array with the _new_ value for `users` and the _old_ value for `pets`.  This works great if users and pets update independently (i.e. seconds apart), but in the use case above they are expected to update as a part of the same action.  *Even though they update from the same action, they will each emit a new value, and combine latest will therefore always emit 2 values, the first of which will have one of the two being the cached value*

This is not a bug in `combineLatest` or Ngrx. *It is what combineLatest does*

The alternative i put forth in gitter is to only select once from the store, but select both needed values:

```typescript
store.select(s => [s.users, s.pets])
```

This will always emit with the most recent value for each of users and pets.  However, because the selector function create's a new array every time, `select`'s internal `distinctUntilChanged` call will not work, because it checks for equality by object reference.  As a result, that selector will emit *every time the state updates*, not just when `users` or `pets` updates.  To prevent this, we need our own `distinctUntilChanged` with a custom comparator, which is what @maxime1992 posted earlier:

```typescript
store.select(s => [s.users, s.pets])
   .distinctUntilChanged((prev, next) => prev[0] === next[0] && prev[1] === next[1])
```

Another solution if one is committed to using `combineLatest` is just to debounce the observable:

```typescript
store.select(s => s.users)
   .combineLatest(store.select(s => s.pets))
   .debounceTime(1)
   .map(([users, pets]) => ...)
```

Because the two emits from combineLatest should happen back to back; we can just use debounceTime to swallow the first of the two emits.


In conclusion, I think this issue is due to using the wrong rx operator rather than some bug in store.

**Comment by maxime1992** [comment]
@adharris I clearly see what you're telling me :)

> why is the selector being triggered after only one reducer has been run

It's not the case, it's just rxjs and the "cache" from combineLatest ! 

Thanks for giving such a detailed explanation.

I think it might be worth to note somewhere that when combining streams, it's a good idea to use that notation `store.select(s => [s.users, s.pets])` instead of `combineLatest` as I did.

Because many people told me to go that way, except that they probably don't have a normalized state and they don't need to link data between different parts of the store.

Thanks @adharris,

I'm closing the issue as it's not a bug related to ngrx (nor a bug at all in fact).

Just poor design from me :)
